File: backend/app/db/demo_fixture.py
"""김민수 데모 데이터의 단일 원본을 읽는다.

사용자 앱의 `user-7d4e9a2c5f18` 와 트레이너 앱의 `seed-client-1` 은 같은 사람이다.
예전에는 두 앱과 백엔드가 각자 알고리즘으로 그의 과거를 만들어서, 같은 날짜를
나란히 놓으면 숫자가 어긋났다(#757). 이제 셋 다 이 픽스처만 읽는다 — 여기에
계산은 없고, 픽스처에 적힌 값을 날짜에 붙이는 일만 한다.

**이행률은 저장하지 않는다.** [FixtureDay.completion] 은 그날 루틴 항목 중
`done` 인 것의 비율이다. 퍼센트를 따로 적어 두면 화면의 "3개 중 2개 완료" 와
숫자가 갈라진다.

Dart 쪽 짝은 `shared/demo_fixture/lib/demo_fixture.dart` 다 — 날짜를 붙이는
규칙(주 격자 + 최근 사흘 덮어쓰기)을 양쪽이 똑같이 구현한다.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import date, timedelta
from functools import lru_cache
from pathlib import Path

#: 픽스처 파일. `shared/demo_fixture/assets/kim_minsu.json` 과 내용이 같다 —
#: 백엔드 이미지가 `backend/` 만 담기 때문에 여기에도 둔다(`tool/gen_demo_fixture.py`
#: 가 두 곳에 함께 쓴다).
FIXTURE_PATH = Path(__file__).with_name("demo_fixture_data.json")

_DAY_LABELS = ("월", "화", "수", "목", "금", "토", "일")


@dataclass(frozen=True)
class FixtureFood:
    name: str
    calories: int
    sodium_mg: int
    sugar_g: float
    carbs_g: float
    protein_g: float
    fat_g: float

    def as_row(self) -> dict:
        """화면·저장소가 읽는 키 이름 그대로. 세 곳이 같은 키를 쓴다."""
        return {
            "name": self.name,
            "calories": self.calories,
            "sodium_mg": self.sodium_mg,
            "sugar_g": self.sugar_g,
            "carbs_g": self.carbs_g,
            "protein_g": self.protein_g,
            "fat_g": self.fat_g,
        }


@dataclass(frozen=True)
class FixtureMeal:
    slug: str
    #: 시연 중 화면에서 지목하는 행만 id 를 못 박는다. 나머지는 None 이고 호출부가
    #: 날짜로 만든다.
    row_id: str | None
    meal_type: str
    time_label: str
    photo_asset: str
    ai_comment: str
    foods: tuple[FixtureFood, ...]

    @property
    def calories(self) -> int:
        return sum(f.calories for f in self.foods)

    @property
    def sodium_mg(self) -> int:
        return sum(f.sodium_mg for f in self.foods)

    @property
    def sugar_g(self) -> float:
        return round(sum(f.sugar_g for f in self.foods), 1)

    @property
    def carbs_g(self) -> float:
        return round(sum(f.carbs_g for f in self.foods), 1)

    @property
    def protein_g(self) -> float:
        return round(sum(f.protein_g for f in self.foods), 1)

    @property
    def fat_g(self) -> float:
        return round(sum(f.fat_g for f in self.foods), 1)

    def foods_json(self) -> str:
        return json.dumps([f.as_row() for f in self.foods], ensure_ascii=False)


@dataclass(frozen=True)
class FixtureExercise:
    name: str
    #: `cardio` | `strength` | `stretching` | `other` — 표준 어휘 (#996, #997).
    type: str
    minutes: int
    calories: int
    done: bool
    #: 근력 항목의 세트 수와 한 세트당 횟수. 다른 유형은 None 이다.
    #:
    #: 예전에는 이 값을 읽지 않고 버렸다 — Dart 쪽 짝은 읽는데 백엔드만 버려서,
    #: 실 API 로 붙이면 같은 회원의 같은 날 근력이 화면마다 다른 세트 수로
    #: 보였다(분에서 되짚은 수와 픽스처가 적어 둔 수가 달랐다). (#1265)
    sets: int | None = None
    reps: int | None = None
    #: 근력 항목의 중량(kg). 맨몸 운동은 0 이다 — 근력이면 언제나 값을 하나
    #: 든다(#1902). 예전에는 이 값이 이름 문자열에만 있어(`레그프레스 70kg`),
    #: 이름을 쓰는 화면과 필드를 읽는 화면이 같은 기록을 다르게 말했다.
    weight: float | None = None

    @property
    def label(self) -> str:
        """트레이너 화면이 쓰는 표기. 이행률과 이 목록이 같은 자리에서 나온다(#754)."""
        return f"{self.name} {'✓' if self.done else '✗'}"


@dataclass(frozen=True)
class FixtureRoutine:
    """회원에게 배정된 개인운동 한 줄. Dart 쪽 `FixtureRoutine` 과 같은 내용이다.

    출처(`source`)가 픽스처에 있는 것이 중요하다 — 회원 앱은 이 값으로 `트레이너
    추천` 과 `AI 추천 · 트레이너 확인` 을 갈라 적는다. 실서버 시드가 따로 든 표를
    읽던 동안은 이 구분이 실연동에서 사라졌다(#1199).
    """

    #: 실서버 시드가 쓰는 것과 같은 id. 두 앱이 같은 루틴을 같은 줄로 가리킨다.
    id: str
    name: str
    minutes: int
    #: 운동 유형 — 유산소|근력|스트레칭|기타 (모델과 같은 한국어 어휘).
    type: str
    reason: str
    #: `ai` | `trainer`.
    source: str
    #: 근력 배정의 세트 수·한 세트당 횟수·중량(kg). 다른 유형은 None 이다(#1276)
    #: — 유형마다 재는 단위가 다르다. 맨몸 운동은 중량만 비어 있다.
    sets: int | None = None
    reps: int | None = None
    weight: float | None = None


@dataclass(frozen=True)
class FixtureDay:
    day: date
    meals: tuple[FixtureMeal, ...]
    exercises: tuple[FixtureExercise, ...]
    routine_label: str
    is_pt: bool
    client_feedback: str
    trainer_note: str
    day_message: str

    @property
    def iso(self) -> str:
        return self.day.isoformat()

    @property
    def week_start(self) -> str:
        """그 주 월요일. 운동은 주 단위로 조회된다."""
        return (self.day - timedelta(days=self.day.weekday())).isoformat()

    @property
    def day_label(self) -> str:
        return _DAY_LABELS[self.day.weekday()]

    @property
    def has_record(self) -> bool:
        return bool(self.meals or self.exercises)

    @property
    def calories(self) -> int:
        return sum(m.calories for m in self.meals)

    @property
    def sodium_mg(self) -> int:
        return sum(m.sodium_mg for m in self.meals)

    @property
    def sugar_g(self) -> float:
        return round(sum(m.sugar_g for m in self.meals), 1)

    @property
    def completion(self) -> int:
        """이행률(%) — 저장된 값이 아니라 `done` 개수에서 나온다. 휴식일은 0."""
        if not self.exercises:
            return 0
        done = sum(1 for e in self.exercises if e.done)
        return round(done * 100 / len(self.exercises))

    @property
    def done_exercises(self) -> tuple[FixtureExercise, ...]:
        """실제로 한 운동만. 주간 활동 합계가 이걸 쌓는다."""
        return tuple(e for e in self.exercises if e.done)


class DemoFixture:
    """픽스처 파일 한 벌."""

    def __init__(self, payload: dict) -> None:
        member = payload["member"]
        self.member_name: str = member["name"]
        #: 사용자 앱 데모 계정 id.
        self.user_app_seed_id: str = member["userAppSeedId"]
        #: 트레이너 앱에서 같은 사람을 가리키는 고객 id.
        self.trainer_client_id: str = member["trainerClientId"]
        #: 픽스처가 덮는 주 수(이번 주 포함).
        self.history_weeks: int = payload["historyWeeks"]
        self._foods = {
            key: FixtureFood(
                name=value["name"],
                calories=value["calories"],
                sodium_mg=value["sodiumMg"],
                sugar_g=value["sugarG"],
                carbs_g=value["carbsG"],
                protein_g=value["proteinG"],
                fat_g=value["fatG"],
            )
            for key, value in payload["foods"].items()
        }
        self._meals: dict[str, dict] = payload["meals"]
        #: 배정된 개인운동. 실서버 시드도 이 목록을 읽는다(#1199).
        self.routines: tuple[FixtureRoutine, ...] = tuple(
            FixtureRoutine(
                id=item["id"],
                name=item["name"],
                minutes=item["minutes"],
                type=item["type"],
                reason=item["reason"],
                source=item["source"],
                sets=item.get("sets"),
                reps=item.get("reps"),
                weight=item.get("weight"),
            )
            for item in payload.get("routines", ())
        )
        self._recent: list[dict] = payload["recent"]
        self._weeks: list[dict] = payload["weeks"]

    def days_for(self, today: date) -> list[FixtureDay]:
        """[today] 기준으로 날짜가 붙은 하루들을 오래된 → 오늘 순으로 돌려준다.

        주 격자가 달력 주를 채우고, 최근 사흘(오늘·어제·그제)이 그 위를 덮는다.
        아직 오지 않은 요일은 넣지 않는다 — 넣으면 주간 추이 그래프가 빈 날을
        막대로 그리고 주 평균도 실제보다 높아진다(#752).
        """
        this_monday = today - timedelta(days=today.weekday())
        by_date: dict[date, FixtureDay] = {}

        for week in self._weeks:
            week_monday = this_monday - timedelta(days=7 * week["weeksAgo"])
            for entry in week["days"]:
                day = week_monday + timedelta(days=entry["weekday"])
                if day > today:
                    continue
                by_date[day] = self._day_from(entry, day)

        for entry in self._recent:
            day = today - timedelta(days=entry["offset"])
            by_date[day] = self._day_from(entry, day)

        return [by_date[key] for key in sorted(by_date)]

    def _day_from(self, entry: dict, day: date) -> FixtureDay:
        return FixtureDay(
            day=day,
            meals=tuple(self._meal_from(meal) for meal in entry["meals"]),
            exercises=tuple(
                FixtureExercise(
                    name=item["name"],
                    type=item["type"],
                    minutes=item["minutes"],
                    calories=item["calories"],
                    done=item["done"],
                    sets=item.get("sets"),
                    reps=item.get("reps"),
                )
                for item in entry["exercises"]
            ),
            routine_label=entry.get("label", ""),
            is_pt=entry.get("pt", False),
            client_feedback=entry.get("clientFeedback", ""),
            trainer_note=entry.get("trainerNote", ""),
            day_message=entry.get("dayMessage", ""),
        )

    def _meal_from(self, entry: dict) -> FixtureMeal:
        slug = entry["meal"]
        template = self._meals[slug]
        return FixtureMeal(
            slug=slug,
            row_id=entry.get("id"),
            meal_type=template["mealType"],
            time_label=entry.get("timeLabel", template["timeLabel"]),
            photo_asset=template["photoAsset"],
            ai_comment=entry.get("aiComment", template["aiComment"]),
            foods=tuple(self._foods[key] for key in template["foods"]),
        )


@lru_cache(maxsize=1)
def load_fixture() -> DemoFixture:
    """픽스처를 읽는다. 파일은 바뀌지 않으므로 한 번만 읽는다."""
    return DemoFixture(json.loads(FIXTURE_PATH.read_text(encoding="utf-8")))

File: shared/demo_fixture/lib/demo_fixture.dart
/// 김민수 데모 데이터의 단일 원본을 읽는다.
///
/// 사용자앱(`user-7d4e9a2c5f18`)과 트레이너앱(`seed-client-1`)은 같은 사람을 보여 준다.
/// 예전에는 두 앱과 백엔드가 각자 알고리즘으로 그의 과거를 만들어서, 같은 날짜를
/// 나란히 놓으면 숫자가 어긋났다(#757). 이제 셋 다 이 픽스처만 읽는다 — 여기에
/// 계산은 없고, 픽스처에 적힌 값을 날짜에 붙이는 일만 한다.
///
/// **이행률은 저장하지 않는다.** `FixtureDay.completion` 은 그날 루틴 항목 중
/// `done` 인 것의 비율이다. 퍼센트를 따로 적어 두면 화면의 "3개 중 2개 완료" 와
/// 숫자가 갈라진다.
library;

import 'dart:convert';

import 'package:demo_fixture/src/fixture_json.g.dart';

/// 김민수의 담당 트레이너 — 데모 트레이너 계정의 이름. (#2062)
///
/// 회원 앱의 목 코치·목 헬스장·데모 알림 문구와 트레이너 웹의 데모 프로필이 모두
/// 이 값을 읽는다. 문구에 이름을 글자로 박지 않고 이 값을 끼워 넣어, 이름을 바꿀
/// 때 한 곳만 고치면 되게 한다. 백엔드 시드의 같은 사람은
/// `backend/app/db/seed_trainer.py` 의 `TRAINER_NAME` 이다 — 둘은 같아야 한다.
///
/// 예전에는 `김트레이너` 라는 자리표시자였다. 이름과 직함을 한 줄에 두면
/// `김트레이너 퍼스널 트레이너` 처럼 `트레이너` 가 두 번 읽혀 실제 이름으로 바꿨다.
const String kDemoTrainerName = '김태오';

/// 음식 한 가지.
class FixtureFood {
  const FixtureFood({
    required this.name,
    required this.calories,
    required this.sodiumMg,
    required this.sugarG,
    required this.carbsG,
    required this.proteinG,
    required this.fatG,
  });

  factory FixtureFood.fromJson(Map<String, Object?> json) => FixtureFood(
    name: json['name']! as String,
    calories: (json['calories']! as num).toInt(),
    sodiumMg: (json['sodiumMg']! as num).toInt(),
    sugarG: (json['sugarG']! as num).toDouble(),
    carbsG: (json['carbsG']! as num).toDouble(),
    proteinG: (json['proteinG']! as num).toDouble(),
    fatG: (json['fatG']! as num).toDouble(),
  );

  final String name;
  final int calories;
  final int sodiumMg;
  final double sugarG;
  final double carbsG;
  final double proteinG;
  final double fatG;

  /// 화면·저장소가 읽는 키 이름 그대로. 세 곳이 같은 키를 쓴다.
  Map<String, Object?> toJson() => <String, Object?>{
    'name': name,
    'calories': calories,
    'sodium_mg': sodiumMg,
    'sugar_g': sugarG,
    'carbs_g': carbsG,
    'protein_g': proteinG,
    'fat_g': fatG,
  };
}

/// 하루에 놓인 끼니 하나. 끼니 템플릿에 그날만의 id·시각·코멘트가 덮인 결과다.
class FixtureMeal {
  const FixtureMeal({
    required this.slug,
    required this.rowId,
    required this.mealType,
    required this.timeLabel,
    required this.photoAsset,
    required this.aiComment,
    required this.foods,
  });

  /// 끼니 템플릿 이름(`lunch-jjamppong` 등).
  final String slug;

  /// 시연 중 화면에서 지목하는 행만 id 를 못 박는다. 나머지는 null 이고, 소비자가
  /// 날짜로 id 를 만든다.
  final String? rowId;

  final String mealType;
  final String timeLabel;
  final String photoAsset;
  final String aiComment;
  final List<FixtureFood> foods;

  int get calories =>
      foods.fold<int>(0, (int sum, FixtureFood f) => sum + f.calories);
  int get sodiumMg =>
      foods.fold<int>(0, (int sum, FixtureFood f) => sum + f.sodiumMg);
  double get sugarG => _round1(
    foods.fold<double>(0, (double sum, FixtureFood f) => sum + f.sugarG),
  );
  double get carbsG => _round1(
    foods.fold<double>(0, (double sum, FixtureFood f) => sum + f.carbsG),
  );
  double get proteinG => _round1(
    foods.fold<double>(0, (double sum, FixtureFood f) => sum + f.proteinG),
  );
  double get fatG => _round1(
    foods.fold<double>(0, (double sum, FixtureFood f) => sum + f.fatG),
  );

  String foodsJson() => jsonEncode(<Map<String, Object?>>[
    for (final FixtureFood f in foods) f.toJson(),
  ]);
}

/// 그날 루틴의 항목 하나.
class FixtureExercise {
  const FixtureExercise({
    required this.name,
    required this.type,
    required this.minutes,
    required this.calories,
    required this.done,
    this.sets,
    this.reps,
    this.weight,
  });

  factory FixtureExercise.fromJson(Map<String, Object?> json) =>
      FixtureExercise(
        name: json['name']! as String,
        type: json['type']! as String,
        minutes: (json['minutes']! as num).toInt(),
        calories: (json['calories']! as num).toInt(),
        done: json['done']! as bool,
        sets: (json['sets'] as num?)?.toInt(),
        reps: (json['reps'] as num?)?.toInt(),
        weight: (json['weight'] as num?)?.toDouble(),
      );

  final String name;

  /// `cardio` | `strength` | `stretching`.
  final String type;
  final int minutes;
  final int calories;
  final bool done;

  /// 근력 항목의 **세트 수**. 이름에 적힌 `4세트` 를 화면이 다시 세거나 분에서
  /// 되짚어 계산하면 세 화면이 저마다 다른 수를 말한다 — 값으로 들고 다닌다.
  /// 유산소·스트레칭·기타는 분이 곧 값이라 null 이다.
  final int? sets;

  /// 근력 항목의 **한 세트당 횟수**. 세트·중량과 한 벌이다(#1310) — 셋이 다
  /// 있어야 지난주에 무엇을 했는지 그대로 되짚는다.
  final int? reps;

  /// 근력 항목의 **중량**(kg). 맨몸 운동은 `0` 이다 — 근력이면 언제나 값을
  /// 하나 든다(#1902). 예전에는 이 값이 이름 문자열에만 있어(`레그프레스 70kg`)
  /// 이름을 쓰는 화면과 필드를 읽는 화면이 같은 기록을 다르게 말했다.
  final double? weight;

  /// 트레이너 화면이 쓰는 표기. 이행률과 이 목록이 같은 자리에서 나오므로
  /// "67%" 옆에 "3개 중 3개 완료" 가 놓이는 일이 없다(#754).
  String get label => '$name ${done ? '✓' : '✗'}';
}

/// 픽스처가 말하는 하루.
class FixtureDay {
  const FixtureDay({
    required this.date,
    required this.weekStart,
    required this.dayLabel,
    required this.meals,
    required this.exercises,
    required this.routineLabel,
    required this.isPt,
    required this.clientFeedback,
    required this.trainerNote,
    required this.dayMessage,
  });

  /// `YYYY-MM-DD`.
  final String date;

  /// 그 주 월요일(`YYYY-MM-DD`). 운동은 주 단위로 조회된다.
  final String weekStart;

  /// 월~일 한 글자.
  final String dayLabel;

  final List<FixtureMeal> meals;
  final List<FixtureExercise> exercises;

  /// `PT 세션 · 트레이너 지도` 같은 종류 라벨.
  final String routineLabel;
  final bool isPt;
  final String clientFeedback;
  final String trainerNote;

  /// 식단 탭의 하루 코치 문구. 큐레이션 사흘만 갖는다.
  final String dayMessage;

  bool get hasRecord => meals.isNotEmpty || exercises.isNotEmpty;

  int get calories =>
      meals.fold<int>(0, (int sum, FixtureMeal m) => sum + m.calories);
  int get sodiumMg =>
      meals.fold<int>(0, (int sum, FixtureMeal m) => sum + m.sodiumMg);
  double get sugarG => _round1(
    meals.fold<double>(0, (double sum, FixtureMeal m) => sum + m.sugarG),
  );

  /// 그날의 탄·단·지(g). 끼니에서 그대로 합친다 — 트레이너 화면의 `이번 달`
  /// 칼로리 막대를 탄단지로 쌓는 재료다(#944). 당류와 같이 소수를 유지한다.
  double get carbsG => _round1(
    meals.fold<double>(0, (double sum, FixtureMeal m) => sum + m.carbsG),
  );
  double get proteinG => _round1(
    meals.fold<double>(0, (double sum, FixtureMeal m) => sum + m.proteinG),
  );
  double get fatG => _round1(
    meals.fold<double>(0, (double sum, FixtureMeal m) => sum + m.fatG),
  );

  /// 이행률(%) — 저장된 값이 아니라 `done` 개수에서 나온다. 루틴이 없는 날(휴식)은
  /// 0 이다.
  int get completion {
    if (exercises.isEmpty) return 0;
    final int done = exercises.where((FixtureExercise e) => e.done).length;
    return (done * 100 / exercises.length).round();
  }

  /// 실제로 한 운동만. 사용자 앱의 주간 활동 그래프가 이걸 쌓는다.
  List<FixtureExercise> get doneExercises =>
      exercises.where((FixtureExercise e) => e.done).toList();
}

/// 김민수에게 배정된 개인 운동 하나. (#1170)
///
/// 회원 앱 `추천 개인운동`, 트레이너 고객 탭 `아직 하지 않은 개인 운동`,
/// 트레이너 프로그램 탭이 **모두 이 목록 하나**를 읽는다. 예전에는 세 화면이
/// 각자 목록을 들고 있어서, 같은 회원의 같은 날에 셋이 서로 다른 운동을 말했다.
class FixtureRoutine {
  /// Creates one assigned routine.
  const FixtureRoutine({
    required this.id,
    required this.name,
    required this.minutes,
    required this.type,
    required this.reason,
    required this.source,
    this.sets,
    this.reps,
    this.weight,
  });

  factory FixtureRoutine.fromJson(Map<String, Object?> json) => FixtureRoutine(
    id: json['id']! as String,
    name: json['name']! as String,
    minutes: (json['minutes']! as num).toInt(),
    type: json['type']! as String,
    reason: json['reason']! as String,
    source: json['source']! as String,
    sets: (json['sets'] as num?)?.toInt(),
    reps: (json['reps'] as num?)?.toInt(),
    weight: (json['weight'] as num?)?.toDouble(),
  );

  /// 실서버 시드가 쓰는 것과 같은 id. 두 앱이 같은 루틴을 같은 이름으로
  /// 가리켜야 완료 표시가 어느 화면에서 눌려도 같은 줄에 붙는다.
  final String id;

  final String name;
  final int minutes;

  /// `유산소` | `근력` | `스트레칭` — 화면이 그대로 쓰는 한국어다.
  final String type;

  final String reason;

  /// `ai` | `trainer` — 누가 정했는지. 회원 화면이 줄마다 출처를 적는다.
  final String source;

  /// 근력 배정의 세트 수·한 세트당 횟수·중량(kg). 다른 유형은 비어 있다(#1276)
  /// — 유형마다 재는 단위가 다르다. 맨몸 운동은 중량만 비어 있다: 적지 않은
  /// 값을 `0kg` 으로 적으면 트레이너가 정해 준 무게처럼 읽힌다.
  final int? sets;
  final int? reps;
  final double? weight;
}

/// 픽스처 파일 한 벌.
class DemoFixture {
  DemoFixture._({
    required this.memberName,
    required this.userAppSeedId,
    required this.trainerClientId,
    required this.historyWeeks,
    required this.routines,
    required Map<String, FixtureFood> foods,
    required Map<String, Map<String, Object?>> meals,
    required List<Map<String, Object?>> recent,
    required List<Map<String, Object?>> weeks,
  }) : _foods = foods,
       _meals = meals,
       _recent = recent,
       _weeks = weeks;

  /// 앱에 심긴 픽스처. **동기**다.
  ///
  /// 에셋으로 두지 않는 이유: 에셋을 읽으려면 `rootBundle` 이 필요하고 그건 비동기라,
  /// 가짜 비동기 안에서 도는 위젯 테스트가 펌프될 때까지 풀리지 않아 통째로 멈춘다.
  /// 시드는 위젯 테스트의 부팅 경로에서도 불리므로 그 함정을 아예 없앤다.
  factory DemoFixture.load() => DemoFixture.parse(kimMinsuFixtureJson);

  /// 주어진 JSON 문자열에서 만든다. 테스트가 원본 파일을 직접 읽어 넣을 때 쓴다 —
  /// 그래야 심긴 상수가 원본과 갈라졌는지도 함께 드러난다.
  factory DemoFixture.parse(String source) {
    final Map<String, Object?> json =
        jsonDecode(source) as Map<String, Object?>;
    final Map<String, Object?> member = json['member']! as Map<String, Object?>;
    return DemoFixture._(
      memberName: member['name']! as String,
      userAppSeedId: member['userAppSeedId']! as String,
      trainerClientId: member['trainerClientId']! as String,
      historyWeeks: (json['historyWeeks']! as num).toInt(),
      routines: <FixtureRoutine>[
        for (final Object? routine
            in (json['routines'] as List<Object?>?) ?? const <Object?>[])
          FixtureRoutine.fromJson(routine! as Map<String, Object?>),
      ],
      foods: <String, FixtureFood>{
        for (final MapEntry<String, Object?> e
            in (json['foods']! as Map<String, Object?>).entries)
          e.key: FixtureFood.fromJson(e.value! as Map<String, Object?>),
      },
      meals: <String, Map<String, Object?>>{
        for (final MapEntry<String, Object?> e
            in (json['meals']! as Map<String, Object?>).entries)
          e.key: e.value! as Map<String, Object?>,
      },
      recent: <Map<String, Object?>>[
        for (final Object? day in json['recent']! as List<Object?>)
          day! as Map<String, Object?>,
      ],
      weeks: <Map<String, Object?>>[
        for (final Object? week in json['weeks']! as List<Object?>)
          week! as Map<String, Object?>,
      ],
    );
  }

  final String memberName;

  /// 사용자 앱 데모 계정 id.
  final String userAppSeedId;

  /// 트레이너 앱에서 같은 사람을 가리키는 고객 id.
  final String trainerClientId;

  /// 픽스처가 덮는 주 수(이번 주 포함).
  final int historyWeeks;

  /// 배정된 개인 운동. 세 화면이 같은 목록을 읽는다(#1170).
  final List<FixtureRoutine> routines;

  final Map<String, FixtureFood> _foods;
  final Map<String, Map<String, Object?>> _meals;
  final List<Map<String, Object?>> _recent;
  final List<Map<String, Object?>> _weeks;

  /// [now] 기준으로 날짜가 붙은 하루들을 오래된 → 오늘 순으로 돌려준다.
  ///
  /// 주 격자가 달력 주를 채우고, 최근 사흘(오늘·어제·그제)이 그 위를 덮는다.
  /// 아직 오지 않은 요일은 넣지 않는다 — 넣으면 주간 추이 그래프가 빈 날을
  /// 막대로 그리고 주 평균도 실제보다 높아진다(#752).
  List<FixtureDay> daysFor(DateTime now) {
    final DateTime today = DateTime(now.year, now.month, now.day);
    final DateTime thisMonday = _addDays(today, -(today.weekday - 1));

    final Map<String, FixtureDay> byDate = <String, FixtureDay>{};
    for (final Map<String, Object?> week in _weeks) {
      final int weeksAgo = (week['weeksAgo']! as num).toInt();
      final DateTime weekMonday = _addDays(thisMonday, -7 * weeksAgo);
      for (final Object? entry in week['days']! as List<Object?>) {
        final Map<String, Object?> day = entry! as Map<String, Object?>;
        final DateTime date = _addDays(
          weekMonday,
          (day['weekday']! as num).toInt(),
        );
        if (date.isAfter(today)) continue;
        byDate[_ymd(date)] = _dayFrom(day, date);
      }
    }

    for (final Map<String, Object?> day in _recent) {
      final DateTime date = _addDays(today, -(day['offset']! as num).toInt());
      byDate[_ymd(date)] = _dayFrom(day, date);
    }

    final List<String> dates = byDate.keys.toList()..sort();
    return <FixtureDay>[for (final String date in dates) byDate[date]!];
  }

  FixtureDay _dayFrom(Map<String, Object?> day, DateTime date) {
    final DateTime monday = _addDays(date, -(date.weekday - 1));
    return FixtureDay(
      date: _ymd(date),
      weekStart: _ymd(monday),
      dayLabel: _dayLabels[date.weekday - 1],
      meals: <FixtureMeal>[
        for (final Object? entry in day['meals']! as List<Object?>)
          _mealFrom(entry! as Map<String, Object?>),
      ],
      exercises: <FixtureExercise>[
        for (final Object? entry in day['exercises']! as List<Object?>)
          FixtureExercise.fromJson(entry! as Map<String, Object?>),
      ],
      routineLabel: day['label'] as String? ?? '',
      isPt: day['pt'] as bool? ?? false,
      clientFeedback: day['clientFeedback'] as String? ?? '',
      trainerNote: day['trainerNote'] as String? ?? '',
      dayMessage: day['dayMessage'] as String? ?? '',
    );
  }

  FixtureMeal _mealFrom(Map<String, Object?> entry) {
    final String slug = entry['meal']! as String;
    final Map<String, Object?> template = _meals[slug]!;
    return FixtureMeal(
      slug: slug,
      rowId: entry['id'] as String?,
      mealType: template['mealType']! as String,
      timeLabel:
          entry['timeLabel'] as String? ?? template['timeLabel']! as String,
      photoAsset: template['photoAsset']! as String,
      aiComment:
          entry['aiComment'] as String? ?? template['aiComment']! as String,
      foods: <FixtureFood>[
        for (final Object? key in template['foods']! as List<Object?>)
          _foods[key! as String]!,
      ],
    );
  }
}

const List<String> _dayLabels = <String>['월', '화', '수', '목', '금', '토', '일'];

/// 날짜만 더한다. `Duration(days:)` 은 서머타임이 있는 지역에서 하루가 23·25시간이
/// 되는 날 날짜가 밀린다 — 데모가 그 날 하루치를 잃는다.
DateTime _addDays(DateTime date, int days) =>
    DateTime(date.year, date.month, date.day + days);

String _ymd(DateTime d) =>
    '${d.year.toString().padLeft(4, '0')}-'
    '${d.month.toString().padLeft(2, '0')}-'
    '${d.day.toString().padLeft(2, '0')}';

double _round1(double value) => (value * 10).round() / 10;
